### PR TITLE
Better shell error handling, add mongo.shell.args setting

### PR DESCRIPTION
**Pull request by StephenWeatherford**

### CODE REVIEW

File: src/mongo/tree/MongoCollectionTreeItem.ts [hunk 1]
```diff
@@ -107,41 +107,36 @@ export class MongoCollectionTreeItem extends AzureParentTreeItem<IMongoTreeRoot>
 		return new MongoDocumentTreeItem(this, newDocument);
 	}
 
-	executeCommand(name: string, args?: string[]): Thenable<string> | null {
+	public async tryExecuteCommandDirectly(name: string, args?: string[]): Promise<{ deferToShell?: boolean; result?: string }> {
```
Comment (StephenWeatherford) [inline src/mongo/tree/MongoCollectionTreeItem.ts]: The original code was really weird and had to understand/debug - it returned a promise if it ran directly, or null (not a promise returning null) if you should run it in the shell instead.

File: test/mongoShell.test.ts [hunk 2]
```diff
@@ -0,0 +1,281 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (StephenWeatherford) [inline test/mongoShell.test.ts]: This is a start at mongo shell tests.  Currently on works on Windows locally (not in pipeline)

File: src/mongo/tree/MongoCollectionTreeItem.ts (hunk 1 above)
Comment (PrashanthCorp) [inline src/mongo/tree/MongoCollectionTreeItem.ts]: Does it need At least one of the fields, exactly one of the fields, or is returning an empty object OK too?

File: src/mongo/tree/MongoDatabaseTreeItem.ts [hunk 3]
```diff
@@ -125,20 +126,33 @@ export class MongoDatabaseTreeItem extends AzureParentTreeItem<IMongoTreeRoot> {
...
+
+		// CONSIDER: Re-using the shell instead of disposing it each time would allow us to keep state
+		//  (JavaScript variables, etc.), but we would need to deal with concurrent requests, or timed-out
+		//  requests.
+		let shell = await this.createShell();
+		try {
+			await shell.useDatabase(this.databaseName);
+			return await shell.executeScript(command.text);
+		} finally {
+			shell.dispose();
```
Comment (PrashanthCorp) [inline src/mongo/tree/MongoDatabaseTreeItem.ts]: Does this mean that everytime a shell is created, we have to ensure we error catch and ensure it is disposed?

Comment (StephenWeatherford) [reply src/mongo/tree/MongoDatabaseTreeItem.ts]: You always want to dispose it.  That's what finally is for.  Without a catch, errors will be bubbled as normal.

File: src/mongo/tree/MongoCollectionTreeItem.ts (hunk 1 above)
Comment (StephenWeatherford) [reply src/mongo/tree/MongoCollectionTreeItem.ts]: Now it's very explicit:  Promise<{ deferToShell: true; result: undefined } | { deferToShell: false; result: string }>

File: src/mongo/MongoShell.ts [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+			},
+			timeoutSeconds * 1000);
+	}
+}
+
+function convertToSingleLine(script: string): string {
+	return script.split(os.EOL)
+		.map(line => line.trim())
+		.join('')
+		.trim();
```
Comment (PrashanthCorp) [inline src/mongo/MongoShell.ts]: Redundant trim?

File: src/mongo/MongoShell.ts [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+}
+
+function removeSentinel(text: string, sentinel: string): { text: string; removed: boolean } {
```
Comment (PrashanthCorp) [inline src/mongo/MongoShell.ts]: Can there be more than one sentinel in the text? This removes just the first occurrence of sentinel.

File: src/mongo/MongoShell.ts [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+					// it back out as a string value after it's done processing the script
+					let quotedSentinel = `"${sentinel}"`;
+					this._process.writeLine(quotedSentinel); // (Don't display the sentinel)
+
+				} catch (error) {
+					// new Promise() doesn't seem to catch exceptions in an async function, we need to explicitly reject it
+
+					if ((<{ code?: string }>error).code === 'EPIPE') {
+						// Give a chance for start-up errors to show up before rejecting with this more general error message
+						await delay(500);
```
Comment (PrashanthCorp) [inline src/mongo/MongoShell.ts]: Does this need to be parameterized?

File: src/utils/InteractiveChildProcess.ts [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+    public static async create(options: IInteractiveChildProcessOptions): Promise<InteractiveChildProcess> {
+        let child: InteractiveChildProcess = new InteractiveChildProcess(options);
+        await child.startCore();
+        return child;
+    }
+
+    public kill(): void {
+        this._isKilling = true;
+        this._childProc.kill();
```
Comment (PrashanthCorp) [inline src/utils/InteractiveChildProcess.ts]: Is this blocking?

File: src/utils/InteractiveChildProcess.ts [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+        this._childProc.stdout.on('data', (data: string | Buffer) => {
+            let text = data.toString();
+            this.writeLineToOutputChannel(text);
+            this._onStdOutEmitter.fire(text);
+        });
+
+        this._childProc.stderr.on('data', (data: string | Buffer) => {
+            let text = data.toString();
+            this._onStdErrEmitter.fire(text);
+            this.writeLineToOutputChannel(text, stdErrPrefix);
```
Comment (PrashanthCorp) [inline src/utils/InteractiveChildProcess.ts]: Why is the order of firing & writing to output different between stdout's and stderr's  handling of data?

File: src/utils/InteractiveChildProcess.ts [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+            while (true) {
+                if (!!this._error || this._isKilling) {
+                    reject(this._error);
+                    break;
+                } else if (!!this._childProc.pid) {
+                    resolve();
+                    break;
+                } else {
+                    if (Date.now() > started + processStartupTimeout) {
+                        reject("The process did not start in a timely manner");
```
Comment (PrashanthCorp) [inline src/utils/InteractiveChildProcess.ts]: Is  there a missing break?

**Pull request merged**